AP_BattMonitor: Move battery parameters to their own class

WickedShell opened the pull request:
This moves all the battery parameters to their own class and allows us to have more then 2 battery monitor instances quite easily. This attempts for the most part to not change any parameter names, however there were 2 main problems with this effort:
1.  `BATT_AMP_PERVOLT` and `BATT2_AMP_PERVOL` unfortunately there wasn't any common names that can be used here that does not result in a parameter name change.
2. `BATT_LOW_TYPE` was renamed to `BATT_FS_VOLTSRC` This is done as my next task is to add a critical battery level as well, and that will be referring to the same parameter so this is future proofing the name.

Point 1 here is the bigger worry as a GCS setup pages are at risk of breaking, but short of hooking in dropping off the last character from the name in AP_Param and adding extra flags for it I couldn't see any way around it. (And looking at it quickly it was not super obvious to me where the AP_Param hook would go, but I'm sure with enough time it could be found).

This does do the appropriate parameter conversions, I've tested them on Plane thus far, and still need to do SITL testing on the other firmware versions (to ensure that I haven't stuffed up the old key for conversions).

Other features of note:
- `_LOW_TIMEOUT` and `_FS_VOLTSRC` are now per batt monitor instance rather then being common to all the batteries, as if you had a separate battery for flight electronics you may want a different set of values there
- I removed several unused helper functions, and killed the weird `#define _BattMonitor_STATE(instance) state[instance]`

amilcarlucas commented on libraries/AP_BattMonitor/AP_BattMonitor_Params.h:
```diff
@@ -0,0 +1,51 @@
+
+protected:
+
+    AP_Int8     _type;             /// 0=disabled, 3=voltage only, 4=voltage and current
```
Nitpicking, whitespace aligment

amilcarlucas commented on libraries/AP_BattMonitor/AP_BattMonitor_Params.h:
```diff
@@ -0,0 +1,51 @@
+    AP_Int16    _watt_max;               /// max battery power allowed. Reduce max throttle to reduce current to satisfy t    his limit
+    AP_Int32    _serial_number;          /// battery serial number, automatically filled in on SMBus batteries
+    AP_Int8     _low_voltage_timeout;    /// timeout in seconds before a low voltage event will be triggered
+    AP_Int8     _failsafe_voltage_source; /// voltage type used for detection of low voltage event
```
whitespace alignment

amilcarlucas commented:
Is rover support missing ? I could not find BattMonitor_* references in sensors.cpp

khancyr commented:
Battery support for rover is very small ...

WickedShell commented:
I glanced at Parameters.h in rover and didn't see it, but I see now it is present, it's just directly made as a GOBJECT in Parameters.cpp. Will update to reflect that.

WickedShell commented:
Param conversion code tested (and fixed the Rover one) in SITL for all Copter/Sub/Rover/Plane. (And Plane was tested in real hardware).

WickedShell commented:
One of the options here is to make `BATT_MONITOR` an enable parameter, this would be quite straightforward and would hide most parameters if/when we move to a 6+ battery system for most users. The key problem with this however is that GCS's expect to be able to show most of the BATT_X parameters on their setup pages, and we currently default to no battery monitor instance which would result in a large GCS headache. I suspect for that reason it doesn't make sense to do that at the moment, but if we wanted to do that, this is the time to do it.

rmackay9 commented:
LGTM.  So we will need to ask michael to change the battery monitor setup pages before the Copter-3.6 release.

amilcarlucas commented:
Can the '_ENABLED' paerameters be used, and default to 'BATT_ENABLED' and 'BATT2_ENABLED' to one and all otheres to zero ?

WickedShell commented:
@amilcarlucas there is no reason to use _ENABLED when you could reused the _TYPE to flag that already. (Type 0 already is a disabled essentially). The only problem with this is the old default values which we would be upgrading are to have the monitors disabled. (And again further breaks the GCS config pages)

WickedShell commented:
`BATT_MONITOR` params are now enable flags with the special case to ensure that the first batt monitor instance is always sent to the GCS to avoid breaking existing setup pages.

magicrub commented on libraries/AP_BattMonitor/AP_BattMonitor_Params.cpp:
```diff
@@ -0,0 +1,92 @@
+    // @Units: W
+    // @Increment: 1
+    // @User: Advanced
+    AP_GROUPINFO_FRAME("WATT_MAX", 8, AP_BattMonitor_Params, _watt_max, 0, AP_PARAM_FRAME_PLANE),
```
are we sure we want to use AP_PARAM_FRAME_PLANE as the default here?

WickedShell replied:
Not following your comment, frame flags are the last thing on a `AP_GROUPINFO_FRAME(name, idx, clazz, element, def, frame_flags)` `WATT_MAX` was plane only before via #define.

magicrub replied:
disregard my comment

magicrub commented:
looks good to me, anything holding this up?

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -227,12 +90,79 @@ AP_BattMonitor::init()
+        {18, (SECOND_BATT_CONVERT_MASK | 7)}, // 2_WATT_MAX
+        {20, (SECOND_BATT_CONVERT_MASK | 8)}, // 2_SERIAL_NUM
+        {21,                             9 }, // _LOW_TIMER
+        {22,                            10 }  // _LOW_TYPE
```
We only had one low timer/type before, now we have 2. The copy to the first one definetly is correct, the more open question is if it should also be copied to the second batt monitor. I guess it probably should (if it was ever set) as that would preserve the behavior the most.

OXINARF replied:
Yes, sorry if I wasn't clear. It is absolutely correct copying to the first one. My doubt is if it should be copied for the second one.
I think my proposal would be to check if a second monitor was configured and, in that case, copy to the second one too. This is a bit more code to do though.

WickedShell commented on libraries/AP_BattMonitor/examples/AP_BattMonitor_test/AP_BattMonitor_test.cpp:
```diff
@@ -18,7 +19,7 @@ void setup() {
 
     // set battery monitor to smbus
-    battery_mon.set_monitoring(0, AP_BattMonitor::BattMonitor_TYPE_SOLO);
+    AP_Param::set_by_name("BATT_MONITOR",  AP_BattMonitor_Params::BattMonitor_TYPE_SOLO);
```
I'm not following this comment at all, could you explain a bit more what you mean?

First of all the prefix is "BATT" from the vehicle (look at plane and sub).

Second I have not gone out of my way to test the examples (I refuse to run them as root, which I'm given to understand means I can't run them) however the call was taken directly from sub 
 https://github.com/ArduPilot/ardupilot/blob/master/ArduSub/system.cpp#L42 and the BATT_MONITOR param will exist.

Side not now that we're talking about it. I somewhat doubt anyone has convincingly run that example, because you can't run SITL/similair with a SMBus driver as there is no fake I2C bus. (I'm not convinced the entire example holds any value but thats a different discussion).

OXINARF replied:
> First of all the prefix is "BATT" from the vehicle (look at plane and sub).

But there is no vehicle in the example... Or better yet, the example is the vehicle, but it has no parameter table.
I'm not sure how else to explain it. When you build an example you aren't building Plane, Copter, Sub, Rover or Tracker, you are building just the example. As you know, the library parameter prefixes are defined in the param tables in the vehicles - there isn't one here, so where does the *BATT_* come from?

> Second I have not gone out of my way to test the examples (I refuse to run them as root, which I'm given to understand means I can't run them) however the call was taken directly from sub
https://github.com/ArduPilot/ardupilot/blob/master/ArduSub/system.cpp#L42 and the BATT_MONITOR param will exist.

Again, you are talking about vehicles that have parameter tables including those parameters. If you go to Sub and change the GND_ to something else (or remove it), then that code you linked to will stop working.

> Side not now that we're talking about it. I somewhat doubt anyone has convincingly run that example, because you can't run SITL/similair with a SMBus driver as there is no fake I2C bus. (I'm not convinced the entire example holds any value but thats a different discussion). (If you set the driver to use a SMBus one in SITL then restart you get a segfault)

You know you can run the examples on a real board right? So, you can run just the example on a Solo for example.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -227,12 +90,79 @@ AP_BattMonitor::init()
+    AP_Param::ConversionInfo info;
+    info.new_name = param_name;
+
+#if APM_BUILD_TYPE(APM_BUILD_ArduPlane)
```
I debated this a bit, but I didn't like the dependency hierarchy. (IE the fact that every vehicle must always ensure that update parameters is called before `init(void)` just felt like a problem waiting to happen in the future.) I ended up preferring the self contained nature of the #define a lot more, and this is one of those things that a vehicle doesn't have any reason to be changing at run time now, and will be forever true until we rebuild the param indexes one day. I can change if people strongly object but I'd prefer to maintain it as is.

OXINARF replied:
I understand your point, but I'm always trying to remove as much vehicle dependencies from libraries as possible. Not only it reduces build dependencies (we only need to compile the file once if it doesn't change based on vehicle), but it also makes it easier to use libraries elsewhere.

As I've written, I would have preferred the other way, but I'll let it go.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_Params.h:
```diff
@@ -0,0 +1,52 @@
+public:
+    static const struct AP_Param::GroupInfo var_info[];
+
+    AP_BattMonitor_Params(void);
```
I'm not comfortable with making this change. I'm not that familiar with overloading C++ constructors, and the interaction with the `create()` methods caused a number of problems when I tried to copy from the GPS example. I'm going to need assistance on making this change if it's a blocking requirement. (Given the partial coverage/current create() situation I'm a bit on the fence how much it's worth worrying about).

WickedShell replied:
Thats what I thought/tried to copy from the GPS library last night, but that will not compile.

```
../../libraries/AP_BattMonitor/AP_BattMonitor.cpp: In constructor ‘AP_BattMonitor::AP_BattMonitor()’:
../../libraries/AP_BattMonitor/AP_BattMonitor.cpp:34:52: error: use of deleted function ‘AP_BattMonitor_Params& AP_BattMonitor_Params::operator=(const AP_BattMonitor_Params&)’
         _params[i] = AP_BattMonitor_Params::create();
                                                    ^
compilation terminated due to -Wfatal-errors.
```

OXINARF replied:
Ok, I know why that happens, the code isn't doing initialization, it's doing copy, which is what we don't want.

Anyway, we have just removed the create methods from master as we found an architectural problem with it. So you can just remove the create method and the for loop and add the non-copy parts, it should work.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_Params.cpp:
```diff
@@ -0,0 +1,92 @@
+    // @User: Advanced
+    AP_GROUPINFO("VOLT_MULT", 4, AP_BattMonitor_Params, _volt_multiplier, AP_BATT_VOLTDIVIDER_DEFAULT),
+
+    // @Param: AMP_PERVOLT
```
Fixed.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -227,12 +90,79 @@ AP_BattMonitor::init()
+    }
+
+    // force _params[0]._type into storage to flag that conversion has been done
+    _params[0]._type.set_and_save(_params[0]._type);
```
Your right, made the change.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -227,12 +90,79 @@ AP_BattMonitor::init()
+#elif APM_BUILD_TYPE(APM_BUILD_ArduSub)
+    info.old_key = 33;
+#elif APM_BUILD_TYPE(APM_BUILD_APMrover2)
+        info.old_key = 145;
```
Fixed.

WickedShell commented on libraries/AP_BattMonitor/AP_BattMonitor_SMBus.cpp:
```diff
@@ -29,12 +30,12 @@ void AP_BattMonitor_SMBus::read(void)
-    if (_full_charge_capacity != _mon._pack_capacity[_state.instance]) {
-        _mon._pack_capacity[_state.instance].set_and_notify(_full_charge_capacity);
+    if (_full_charge_capacity != _params._pack_capacity) {
+        _params._pack_capacity = _full_charge_capacity;
```
Was an accident, fixed.

OXINARF commented on libraries/AP_BattMonitor/AP_BattMonitor_Params.cpp:
```diff
@@ -0,0 +1,92 @@
+
+    // @Param: VOLT_PIN
+    // @DisplayName: Battery Voltage sensing pin
+    // @Description: Setting this to 0 ~ 13 will enable battery voltage sensing on pins A0 ~ A13. On the PX4-v1 it should be set to 100. On the Pixha    wk, Pixracer and NAVIO boards it should be set to 2, Pixhawk2 Power2 is 13.
```
Forgot this one?

WickedShell commented:
Removed `AP_BattMonitor_Params::create()` interfaces, added copy constructors, reverted the removal of the example only `AP_BattMonitor::set_monitoring()`, and squashed commits into a more mergable set.

@OXINARF I believe this is all the review changes, and is ready to be rereviewed. Please don't merge quite yet, as I'd actually like to retest the param upgrade path in SITL one more time after the rebasing just to ensure I haven't broken the conversion code somehow.

OXINARF commented on libraries/AP_BattMonitor/AP_BattMonitor_Params.h:
```diff
@@ -0,0 +1,56 @@
+
+    AP_BattMonitor_Params(void);
+
+    constexpr AP_BattMonitor_Params(AP_BattMonitor_Params &&other) = default;
```
I think it doesn't hurt, but this has also been removed from the other classes when the create was removed.

WickedShell replied:
Addressed.

OXINARF commented on libraries/AP_BattMonitor/AP_BattMonitor_Params.h:
```diff
@@ -0,0 +1,56 @@
+    friend class AP_BattMonitor_Analog;
+    friend class AP_BattMonitor_Backend;
+    friend class AP_BattMonitor_Bebop;
+    friend class AP_BattMonitor_SMBus;
```
Since this AP_BattMonitor_Params is supposed to be used only in AP_BattMonitor, can't we make the AP_x variables public and remove all this friend classes? It avoids having to add a new friend class when a new driver is developed.

WickedShell replied:
It can be, what I was trying to prevent was anyone outside of AP_BattMonitor touching these parameters, however I don't currently see any way for them to get a reference to the `AP_BattMonitor_Param` objects so I think it's safe and have made the change.

WickedShell commented:
Tested param upgrade path on plane on plane with no issues.

OXINARF commented on libraries/AP_BattMonitor/AP_BattMonitor.cpp:
```diff
@@ -227,12 +86,81 @@ AP_BattMonitor::init()
+        {21,                             9 }, // _LOW_TIMER
+        {22,                            10 }, // _LOW_TYPE
+        {21, (SECOND_BATT_CONVERT_MASK | 9)}, // 2_LOW_TIMER
+        {22, (SECOND_BATT_CONVERT_MASK |10)}, // 2_LOW_TYPE
```
Humm so you are always converting these two, I don't quite like that. My idea was that these would only be converted if 2_MONITOR was set to some driver.
If you want to keep as is, let's take this to next dev call to hear others' opinion.

WickedShell replied:
Why not always convert them? Always converting them preserves the existing behavior of the system, which is a good reason to do it IMO.

muramura commented on libraries/AP_BattMonitor/AP_BattMonitor_Analog.cpp:
```diff
@@ -7,11 +7,13 @@
-    _volt_pin_analog_source = hal.analogin->channel(mon._volt_pin[_state.instance]);
-    _curr_pin_analog_source = hal.analogin->channel(mon._curr_pin[_state.instance]);
+    _volt_pin_analog_source = hal.analogin->channel(_params._volt_pin);
+    _curr_pin_analog_source = hal.analogin->channel(_params._volt_pin);
```
I also use "_volt_pin" in "_curr_ pin_analog_source". I think "_curr_ pin".

WickedShell replied:
Very good catch, that was a bad use of copy paste, fixed and pushed.

OXINARF commented:
Finally merged! Thanks!
